### PR TITLE
make ComponentRenderer shouldUpdateComponent return true when props c…

### PR DESCRIPTION
**Pull request by calcsam**
…hange

Fixes https://github.com/gatsbyjs/gatsby/issues/3412

I added a timer in the DefaultLayout of www (similar to what @busticated did in his example gist ^) and this code took 1.5ms (Chrome on a 2yr old MacBook) to execute per prop update. Is not called if Layout doesn't send diff props.

### CODE REVIEW

**Comment by calcsam** [comment]
Also should fix https://github.com/gatsbyjs/gatsby/issues/2074 @cyrilepinat

File: packages/gatsby/cache-dir/component-renderer.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ import PropTypes from "prop-types"
 import loader, { publicLoader } from "./loader"
 import emitter from "./emitter"
 import { apiRunner } from "./api-runner-browser"
+import _ from "lodash"
```
Comment (busticated) [inline packages/gatsby/cache-dir/component-renderer.js]: i'm not sure what the current preference is within the codebase but i usually prefer to pull in individual methods - e.g.

    import isEqual from 'lodash/isEqual'

see "installation" [here](https://github.com/lodash/lodash#installation)

File: packages/gatsby/cache-dir/component-renderer.js [hunk 2]
```diff
@@ -117,6 +118,24 @@ class ComponentRenderer extends React.Component {
       return true
     }
 
+    if (Object.keys(this.props).length !== Object.keys(nextProps).length) {
```
Comment (busticated) [inline packages/gatsby/cache-dir/component-renderer.js]: minor thing but it couldn't hurt to assign to a const here - e.g.

    const propsKeyCount = Object.keys(this.props).length;
    
    if (propsKeyCount !== Object.keys(nextProps).length) { ... }

**Comment by sheedy** [comment]
YES! Thank you @calcsam! I thought not being able to update props in a template from a layout was a limitation of Gatsby. 🎉

**Comment by KyleAMathews** [comment]
I think we should just use this library for checking if props have changed https://github.com/tkh44/shallow-compare It's super small and designed to do exactly what we want.

@sheedy definitely not intended as a limitation 😅

**Comment by calcsam** [comment]
Tested shallow-compare and it fixes the test case identified. It seems to do what we want it to do -- my concern was with equality of object children of props or state but that seems to be addressed here. Going to leave the other stuff in the fn for now -- we can rip it out if no longer necessary.

**Comment by KyleAMathews** [comment]
@calcsam thanks! Publishing new release with this fix.

**Pull request merged**